PR: https://github.com/odoo/odoo/pull/234566

From: e9ddfd0b07f56ad584e66ddc32954d159bd4e093
From: Thibault Delavallée
Date: 2025-11-21 12:29:47

Breaking data model changes scores: del:4 + add:2, change matches:
+    ended_poll_ids = fields.One2many('mail.poll', 'end_message_id')
+    started_poll_ids = fields.One2many('mail.poll', 'start_message_id')
-    ended_poll_ids = fields.One2many('mail.poll', 'end_message_id')
-    started_poll_ids = fields.One2many('mail.poll', 'start_message_id')

Total Changes: 22

[FIX] mail: keep only relevant message types

'mail_poll' is not a message_type. Message types are actually linked
to the source, and have a meaning about expected recipients as well
as follower management. We currently have
 -> generic types depending on global source
 * comment: generated by user (Send a Message, Log a Note);
 * notification: generated automatically during the business flow;
 * email: generated by mail gateway, incoming email;
 * email_outgoing: generated by mass mailing, outgoing email, and
   generally targets only specific recipients while skipping followers
   and notification process;
 -> specific types for recipients / access management
 * user_notification: generated in some business flows, restricted to
   specific recipients (others cannot access);
 * auto_comment: automated answer generated generally by tracking
   associated to templates, which denotates that we want to include
   specific recipients while skipping followers (e.g. acknowledgement
   emails);
 * out_of_office: generated due to an out-of-office mechanism, which
   means specific recipients, and specific management to avoid loops
   and spam;
 -> other notification mechanisms
 * sms: generated by SMS module, denotates a log of an outgoing SMS;
 * whatsapp: generated by Whatsap module, denotates a log of an
   outgoing Whatsapp message;

Polls have no specific behavior linked to them. Those are actually
comments with some specific content, which does not require a message
type.

This also allows to cleanup a bit message types, as a type linked to
tracking (content and recipients specific management also) is about
to be added.

Left for upcoming work: investigate whether storing a flag in body
would allow to remove some extra queries linked to reading o2m for
polls. Could also offer a solution for removing 'out of office'
message_type.

Task-5242100
Prepares Task-3645865

closes odoo/odoo#234566

Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/mail/models/mail_message.py
+++ b/addons/mail/models/mail_message.py
@@ -125,7 +125,6 @@ class MailMessage(models.Model):
         ('auto_comment', 'Automated Targeted Notification'),
         ('out_of_office', 'Out-of-office Message'),
         ('user_notification', 'User Specific Notification'),
-        ('mail_poll', 'Poll')
         ],
         'Type', required=True, default='comment',
         help="Used to categorize message generator"
@@ -135,7 +134,6 @@ class MailMessage(models.Model):
              "\n'notification': generated by system e.g. tracking messages"
              "\n'auto_comment': generated by automated notification mechanism e.g. acknowledgment"
              "\n'user_notification': generated for a specific recipient"
-             "\n'mail_poll': generated for a poll"
         )
     subtype_id = fields.Many2one('mail.message.subtype', 'Subtype', ondelete='set null', index=True)
     mail_activity_type_id = fields.Many2one(
@@ -185,6 +183,10 @@ class MailMessage(models.Model):
         groups="base.group_system",
         help='Tracked values are stored in a separate model. This field allow to reconstruct '
              'the tracking and to generate statistics on the model.')
+    # polls
+    ended_poll_ids = fields.One2many('mail.poll', 'end_message_id')
+    started_poll_ids = fields.One2many('mail.poll', 'start_message_id')
+    has_poll = fields.Boolean(compute='_compute_has_poll', compute_sudo=True)
     # mail gateway
     reply_to_force_new = fields.Boolean(
         'No threading for answers',
@@ -203,8 +205,6 @@ class MailMessage(models.Model):
     # as the cache value for this inverse one2many is up-to-date.
     # Besides for new messages, and messages never sending emails, there was no mail, and it was searching for nothing.
     mail_ids = fields.One2many('mail.mail', 'mail_message_id', string='Mails', groups="base.group_system")
-    ended_poll_ids = fields.One2many('mail.poll', 'end_message_id')
-    started_poll_ids = fields.One2many('mail.poll', 'start_message_id')
 
     _model_res_id_idx = models.Index("(model, res_id)")
     _model_res_id_id_idx = models.Index("(model, res_id, id)")
@@ -313,6 +313,10 @@ class MailMessage(models.Model):
             return NotImplemented
         return [('starred_partner_ids', 'in', self.env.user.partner_id.ids)]
 
+    @api.depends('started_poll_ids', 'ended_poll_ids')
+    def _compute_has_poll(self):
+        for message in self:
+            message.has_poll = message.started_poll_ids or message.ended_poll_ids
     # ------------------------------------------------------
     # CRUD / ORM
     # ------------------------------------------------------
@@ -1091,10 +1095,10 @@ class MailMessage(models.Model):
             Store.Many(
                 "started_poll_ids",
                 fields=self.env["mail.poll"]._to_store_defaults(target, with_start_message_id=False),
-                predicate=lambda m: m.message_type == "mail_poll",
+                predicate=lambda m: m.has_poll,
                 sudo=True,
             ),
-            Store.Many("ended_poll_ids", predicate=lambda m: m.message_type == "mail_poll", sudo=True),
+            Store.Many("ended_poll_ids", predicate=lambda m: m.has_poll, sudo=True),
             "subject",
             # sudo: mail.message.subtype - reading subtype on accessible message is allowed
             Store.One("subtype_id", ["description"], sudo=True),
@@ -1365,9 +1369,7 @@ class MailMessage(models.Model):
                 self._has_field_access(self._fields["tracking_value_ids"], "read")
                 and self.tracking_value_ids
             )
-            and not (
-                self.message_type == "mail_poll" and (self.started_poll_ids or self.ended_poll_ids)
-            )
+            and not self.has_poll
         )
 
     @api.model

--- a/addons/mail/models/mail_poll.py
+++ b/addons/mail/models/mail_poll.py
@@ -58,7 +58,7 @@ class MailPoll(models.Model):
         for poll in self:
             thread = thread_by_message[poll.start_message_id]
             poll.end_message_id = thread.message_post(
-                body="", message_type="mail_poll", subtype_xmlid="mail.mt_comment"
+                body="", message_type="comment", subtype_xmlid="mail.mt_comment",
             )
             Store(**thread._get_store_target()).add(poll).bus_send()
 
